src/common/WebView.js

```diff
@@ -0,0 +1,165 @@
+/* @flow strict-local */
```

Comment:
f1a7acaa1 webview: Refactor core logic into a common component.

Cool! It's great to see this core logic centralized so we don't have to duplicate it.

With the choice to do it in a new React component, one thing to be aware of (even if nothing actually changes, and it turns out to be fine) is that it adds a layer of complexity if we want to change the behavior (or debug something) around React components' lifecycles, particularly updates, a.k.a. re-`render`s. I mentioned this briefly [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Resource.20URLs/near/926743) but didn't go into very much detail.

The `WebView` component from `react-native-webview` has its own ideas about when it should update; it can choose to be a `PureComponent` or a regular `Component`, and, if the latter, it can give hints about its desired behavior in `shouldComponentUpdate`.

There's now another layer between that `WebView` and our `MessageList` (and between that `WebView` and the new `EditHistory`, but I think it's much more important to think about `MessageList`).

Our `MessageList` component implements `shouldComponentUpdate` to say that we just never want the `MessageList` component to update. (React would probably say this is controversial; they recommend, "Do not rely on it to “prevent” a rendering, as this can lead to bugs", in [this doc](https://reactjs.org/docs/react-component.html#shouldcomponentupdate). But so far we haven't run into problems with this.)

I haven't tested, but I think everything should actually turn out fine. I would be concerned if our new `WebView` component did some additional updates of its own, not prompted by an update in its parent. The only case I can think of where this would likely happen is if our new `WebView` component had some `state` that it decided, on its own, to change. ("Note that returning `false` does not prevent child components from re-rendering when `their` state changes," from the [`shouldComponentUpdate` doc](https://reactjs.org/docs/react-component.html#shouldcomponentupdate).)

One (optional) simple thing that could be done about this case, for a bit of peace of mind, I think, is to make our new `WebView` a stateless "function component" ([doc](https://reactjs.org/docs/components-and-props.html#function-and-class-components)) and add a note that we shouldn't introduce `state` without considering the extra complexity it would add around re-`render`ing.

Reply:
I also think, just in case we do eventually have to do some hairy debugging around components' lifecycle methods, or just to be clear about what's what, in general, it would be nice to name our new `WebView` something like `ZulipWebView`. We have this kind of name for, e.g., `ZulipButton`, which wraps a component called `Button`.